### PR TITLE
Improve autocomplete

### LINKED ISSUE #4770: [Bug] Improve autocomplete suggestions
## Description
Autocomplete has a "best match" section but many suggestions in that section are not useful.

- [x] API response headers are suggested for table data inputs
- [x] Deeply nested children are shown according to the expected data type, we should limit to 2 child nodes only
- [ ] Ranking needs to improve: Action nodes should be ranked higher for widget autocomplete and Widget nodes should be higher for Action autocomplete
- [ ] Only show "best-match" when no input is added. Once something is typed, default to normal autocomplete
- [x] Suggestions going to multiple lines breaks the styling
- [x] Autocomplete popup should have a max-width

### PR DESCRIPTION
**Pull request by ApekshaBhosale**
## Description
small improvements for autocomplete 

Fixes #4770

## Type of change
- Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>

    // Code coverage diff between base branch:release and head branch: improve-autocomplete 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 51.59 **(0.01)** | 32.66 **(0.01)** | 29.74 **(0)** | 52.11 **(0.01)**
 :green_circle: | app/client/src/components/editorComponents/CodeEditor/index.tsx | 74.52 **(0.25)** | 51.28 **(-0.02)** | 58.06 **(0)** | 75.38 **(0.25)**
 :green_circle: | app/client/src/utils/autocomplete/TernServer.ts | 45.23 **(0.45)** | 31.07 **(1.07)** | 31.48 **(0)** | 50.16 **(0.32)**
 :green_circle: | app/client/src/utils/autocomplete/dataTreeTypeDefCreator.ts | 67.47 **(5.1)** | 61.54 **(-6.32)** | 100 **(0)** | 72.13 **(2.43)**</details>

### CODE REVIEW

File: app/client/src/components/editorComponents/CodeEditor/EditorConfig.ts [hunk 1]
```diff
@@ -46,7 +46,11 @@ export type HintHelper = (
   additionalData?: Record<string, Record<string, unknown>>,
 ) => Hinter;
 export type Hinter = {
-  showHint: (editor: CodeMirror.Editor, expected: string) => void;
+  showHint: (
+    editor: CodeMirror.Editor,
+    expected: string,
+    entity: string,
```
Comment (hetunandu) [inline app/client/src/components/editorComponents/CodeEditor/EditorConfig.ts:52]: rename to entityName

File: app/client/src/utils/autocomplete/TernServer.ts [hunk 2]
```diff
@@ -292,7 +297,12 @@ class TernServer {
 
   getExpectedDataType() {
     const type = this.expected;
-    if (type === "Array<Object>" || type === "Array") return "ARRAY";
+    if (
+      type === "Array<Object>" ||
+      type === "Array" ||
+      type === "Array<{ label: string, value: string }>"
```
Comment (hetunandu) [inline app/client/src/utils/autocomplete/TernServer.ts:303]: Does not handle `Array<x:string, y:number>` for chart data type.
This method does not seem scalable

File: app/client/src/utils/autocomplete/TernServer.ts [hunk 3]
```diff
@@ -152,15 +154,18 @@ class TernServer {
...
       if (data.guess) className += " " + cls + "guess";
-      completions.push({
-        text: completion.name + after,
-        displayText: completion.displayName || completion.name,
-        className: className,
-        data: completion,
-        origin: completion.origin,
-        type: dataType,
-      });
+      if (!entity || !completion.name.includes(entity)) {
```
Comment (hetunandu) [inline app/client/src/utils/autocomplete/TernServer.ts:159]: This check should only be applied when the input is empty. Once the user has written say {{Api1.}} in a API1 field, we should show the correct results then.

File: app/client/src/components/editorComponents/CodeEditor/EditorConfig.ts (hunk 1 above)
Comment (hetunandu) [reply app/client/src/components/editorComponents/CodeEditor/EditorConfig.ts:52]: Also this is optional

**Review by hetunandu — CHANGES_REQUESTED** [review]
Minor changes

File: app/client/src/utils/autocomplete/TernServer.ts (hunk 2 above)
Comment (ApekshaBhosale) [reply app/client/src/utils/autocomplete/TernServer.ts:303]: i will add Array<x:string, y:number> for now we need to change this anyways to take into array<object>

File: app/client/src/utils/autocomplete/TernServer.ts (hunk 3 above)
Comment (ApekshaBhosale) [reply app/client/src/utils/autocomplete/TernServer.ts:159]: this is to remove cyclic dependency we won't show Api1 on Api1

**Review by hetunandu — APPROVED** [review]

File: app/client/src/pages/Editor/APIEditor/Form.tsx [hunk 4]
```diff
@@ -495,6 +495,7 @@ function ApiEditorForm(props: Props) {
           />
           <DatasourceWrapper className="t--dataSourceField">
             <EmbeddedDatasourcePathField
+              dataTreePath={`${actionName}.config.url`}
```
Comment (hetunandu) [inline app/client/src/pages/Editor/APIEditor/Form.tsx:498]:
```suggestion

```

**Review by hetunandu — APPROVED** [review]

**Review by hetunandu — APPROVED** [review]

**Pull request merged**